GGRC-41 Add unsubscribe link to daily digest emails

plamut opened the pull request:
This PR adds an "unsubscribe" link to the bottom of daily digest emails. By clicking the link, users can unsubscribe themselves from receiving daily digest emails.

Things left to do:
- [x] Clean up the code
- [x] Add integration tests for the new API endpoint
- [ ] Fix the "login as user" problem in integration tests
- [x] If a user is not logged in when clicking the link, he should be redirected to the login page first.
  Might already work? Have some trouble reproducing it locally)
- [x] If a user does not have an entry in the `notification_configs` DB table, unsubscribing does not work (because by default, users _are_ subscribed unless there is an entry in the aforementioned table that explicitly override that)
---

Questions:
- ~~Where should the `unsubscribe_from_notifications` be best placed in? Or is having it in `notifications.common` just fine?~~
- ~~How should the API endpoint look like, is having it under `_notifications/` fine?~~
- ~~Do we have any common error handlers for "not authorized", "internal server error", etc. scenarios? Right now those response messages and error codes are manually set.~~
- ~~Simply invoking the API with an email other than user@example.com results in 9 DB queries, even if the handler method body is completely commented out. Is this normal or could e.g. the API path definition be declared in a better way?
(mind that login must be required)~~
- ~~Should we log any attempts to unsubscribe a user other than the currently logged in user himself?~~

-----------------------

**Steps to verify:**
- Make sure a user has at least one notification pending, e.g. assign him/her to an Assessment and modify the latter
- Visit the daily digest page `_notifications/show_daily_digest`
- Find that user's name/email on the page and click the unsubscribe link at the bottom of the email.

**Expected result:**
A user should get unsubscribed. One can verify this by visiting the main application page and opening the user settings menu (top right corner). The _"Daily email digest"_ checkbox should be unchecked. Also, if refreshing the email notifications page, that user's email should not be there anymore.

**Additional notes:**
- When unsubscribing, users must be logged in in the application, and they can only unsubscribe themselves (their own email).

Smotko commented:
> Where should the unsubscribe_from_notifications be best placed in? Or is having it in notifications.common just fine?

notifications/unsubscribe.py?

> How should the API endpoint look like, is having it under _notifications/ fine?

_something is supposed to be a "hidden" url. I think we should simply have `/unsubscribe`.

> Do we have any common error handlers for "not authorized", "internal server error", etc. scenarios? Right now those response messages and error codes are manually set.

No common error handlers yet, but we do have a ticket somewhere in jira to at least make the error pages a bit nicer.

> Simply invoking the API with an email other than user@example.com results in 9 DB queries, even if the handler method body is completely commented out. Is this normal or could e.g. the API path definition be declared in a better way?
(mind that login must be required)

Can you check what those queries are? I'm guessing they are getting the user from the db and loading permissions. Loading permissions might not be necessary for this endpoint though, but I don't think we should worry too much about this - it's not going to be a high traffic endpoint.

> Should we log any attempts to unsubscribe a user other than the currently logged in user himself?

You can do a `logger.warn` in case.

plamut commented:
@Smotko For the reference, I added the changes discussed above, found and fixed an edge case issue, but still have trouble with tests failing do to an uncooperating API helper.

Smotko commented:
@plamut this PR did not see any activity for almost 1 month. I am closing it, but please reopen once you pick it up again.

(Sorry if this is an inconvenience, but I do want to clear the PR queue a bit)
